File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

var binary=function(nums,target,isStart){
    let first =0,last=nums.length-1,ans=-1;
    while(first<=last){
        let mid=Math.floor(first+(last-first)/2);
        if(nums[mid]==target){
            ans=mid;
            if(isStart) last=mid-1;
            else first =mid+1;
    }
        else if(nums[mid]<target) first=mid+1;
        else last=mid-1;
    }
      return ans;
} 
var searchRange = function(nums, target) {
    let start = binary(nums,target,true)
    let end=binary(nums,target,false)
    return [start,end];
   
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int firstOcc(vector<int>& nums, int target){
        int start=0,end=nums.size()-1;
        int ans=-1;
        while(start<=end){
            int mid=start+(end-start)/2;
            if(nums[mid]==target){
                ans=mid;
                end=mid-1;
            } else if(nums[mid]>target){
                end=mid-1;
            }else start=mid+1;     
        } return ans;
    }
    int lastOcc(vector<int>& nums, int target){
        int start=0,end=nums.size()-1;
        int ans=-1;
        while(start<=end){
            int mid=start+(end-start)/2;
            if(nums[mid]==target){
                ans=mid;
                start=mid+1;
            } else if(nums[mid]>target){
                end=mid-1;
            }else start=mid+1;     
        } return ans;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int>ans;
        int firstt=firstOcc(nums,target);
        int lastt=lastOcc(nums,target);
        ans.push_back(firstt);
        ans.push_back(lastt);
        return ans;
    }
};